### ISSUE TITLE
JSDoc should be positioned above all overloads when using TypeScript function overloading

### ISSUE THREAD

**Issue by PindaPixel** [gajus/eslint-plugin-jsdoc#1189]
## Expected behavior

With the following snippet the TypeScript language service correctly displays the JSDoc when inspecting the function, and I expect `jsdoc/require-jsdoc` to know the function is documented.

```ts
/**
 * @param a Some description
 * @param b Some other description
 * @returns The sum of a and b
 */
function add(a:string, b:string):string;
function add(a:number, b:number): number;
function add(a: any, b:any): any {
    return a + b;
}
```

## Actual behavior

The `require-doc` rule reports that the function is missing a JSDoc comment and expects the JSDoc to be positioned under the overloads, but this way the TypeScript language service does not display the given docs upon inspection.

```ts
function add(a:string, b:string):string;
function add(a:number, b:number): number;
/**
 * @param a Some description
 * @param b Some other description
 * @returns The sum of a and b
 */
function add(a: any, b:any): any {
    return a + b;
}
```

See TypeScript language service behaviour [in this example](https://www.typescriptlang.org/play?#code/PQKhFgCgAIWgBADgQwE7ILbWdAygewwFNoATIgZwGNUBLRAF1vwDspYEV0sAjPQkvgYALIqjKUa9Jq3Zx4qIgwCuqFhWgAVUdArKs+AGbZsLUtB5zgUQ8pZUZLbKVIAKZAC4KDOiwDmADQWXj60-gCUIb5+ANw2dg7MTsgu7h4s+jxiQTzpmWKR0BkYWahxkLb2js5unqYAnjkeyCz1hS310ADe7NB9iipqJgDUFuUAvlDxVUk1AExp3tFNS2F+kav+5ZWJrPNpxaVNhwUeRfllUKAQMPJcmCYExBLUdIxJcpxoD3xPgiJiF5Sd6yW4IAaqdRaHR6AzGHAtcyWW7WCoJaopUgLOodJoddqtbq9fpKSEjMZQcZAA).

## ESLint Config

```js
//eslint.config.js
import jsdoc from "eslint-plugin-jsdoc";

export default [jsdoc.configs["flat/recommended-typescript"]];
```

## Environment

- Node version: 20.x
- ESLint version 8.56.0
- `eslint-plugin-jsdoc` version: 48.0.2

**Comment by brettz9** [comment]
There should ideally be some better defaulting behavior here, but currently you can workaround this, by adding a custom rule definition for `jsdoc/require-jsdoc`:

```js
'jsdoc/require-jsdoc': ['warn', {
  contexts: [
    // Expect JSDoc for TypeScript overloads which are not preceded by another overload
    'TSDeclareFunction:not(TSDeclareFunction + TSDeclareFunction)',
    // Expect JSDoc for regular function declarations which are not preceded by an overload
    'FunctionDeclaration:not(TSDeclareFunction + FunctionDeclaration)',
  ],
  require: {
    // Disable the default Function declaration behavior (which applies regardless of preceding overload)
    FunctionDeclaration: false,
  },
}]
```

Ideally, we'd set up any other reasonable contexts for TypeScript, but besides for other cases like this, we don't want to add default requirements beyond looking for JSDoc blocks above function types since projects may differ about the constructs for which they desire JSDoc.

**Comment by PindaPixel** [comment]
Thanks a lot, this is good enough for me. Only downside is that if overload signatures don't match `jsdoc/check-param-names` will report an error `@param "parametername" does not match an existing function parameter`.
